src/snmalloc/backend_helpers/defaultpagemapentry.h

```diff
@@ -43,22 +44,96 @@ namespace snmalloc
      * metadata in meta entries when they are first constructed.
      */
     SNMALLOC_FAST_PATH
-    DefaultPagemapEntry(FrontendSlabMetadata* meta, uintptr_t ras)
-    : FrontendMetaEntry<FrontendSlabMetadata>(meta, ras)
+    DefaultPagemapEntryT(SlabMetadata* meta, uintptr_t ras)
+    : FrontendMetaEntry<SlabMetadata>(meta, ras)
     {}
 
     /**
      * Copy assignment is used only by the pagemap.
      */
-    DefaultPagemapEntry& operator=(const DefaultPagemapEntry& other)
+    DefaultPagemapEntryT& operator=(const DefaultPagemapEntryT& other)
     {
-      FrontendMetaEntry<FrontendSlabMetadata>::operator=(other);
+      FrontendMetaEntry<SlabMetadata>::operator=(other);
       return *this;
     }
 
     /**
      * Default constructor.  This must be callable from the pagemap.
      */
-    SNMALLOC_FAST_PATH DefaultPagemapEntry() = default;
+    SNMALLOC_FAST_PATH DefaultPagemapEntryT() = default;
   };
-} // namespace snmalloc
\ No newline at end of file
+
+  /**
+   * In CHERI, we must retain, internal to the allocator, the authority to
+   * entire backing arenas, as there is no architectural mechanism to splice
+   * together two capabilities.  Additionally, these capabilities will retain
+   * the VMAP software permission, conveying our authority to manipulate the
+   * address space mappings for said arenas.
+   *
+   * We stash these pointers inside the SlabMetadata structures for parts of
+   * the address space for which SlabMetadata exists.  (In other parts of the
+   * system, we will stash them directly in the pagemap.)  This requires that
+   * we inherit from the FrontendSlabMetadata.
+   */
+
+  class StrictProvenanceBackendSlabMetadata : public FrontendSlabMetadata
+  {
+    template<
+      SNMALLOC_CONCEPT(ConceptPAL) A1,
+      typename A2,
+      typename A3,
+      typename A4>
+    friend class BackendAllocator;
+
+    capptr::Arena<void> arena;
+
+    /* Set the arena pointer */
+    void arena_set(capptr::Arena<void> a)
+    {
+      arena = a;
+    }
```

Comment:
Why does this exist, given that it updates a field that has the same visibility as the set method?

Edit: I see that it's necessary to match the interface shared with the Lax version.  Can we have a concept that encapsulates this and some comments explaining why the method exists here?

Reply:
I suppose the arena field could be private?

Reply:
It's a `class`, so it's all private, and the `Backend` is `friend`-ed to allow access.  ("Ah, C++.")

I've moved this out to `src/snmalloc/backend_helpers/cheri_slabmetadata_mixin.h` and added a `concept`.  See https://github.com/microsoft/snmalloc/pull/537/commits/bdb64a25332a1cea54b9eb2fcabc1891bd3c2c08 (which replaces https://github.com/microsoft/snmalloc/commit/72b47b207def5bc12cd39a2b796c28044797ae10).